README.md
# Agouti [![Build Status](https://travis-ci.org/CWISoftware/agouti.png?branch=master)](https://travis-ci.org/CWISoftware/agouti) [![Code Climate](https://codeclimate.com/github/CWISoftware/agouti/badges/gpa.svg)](https://codeclimate.com/github/CWISoftware/agouti) [![Gem Version](https://badge.fury.io/rb/agouti.svg)](http://badge.fury.io/rb/agouti)

Gem for testing above the fold render on the first tcp round trip.

This gem is a Rack middleware that truncates the gzipped response to 14kb.

Useful for testing [critical rendering path optimization](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'agouti'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install agouti

## Usage

To enable the middleware, it is necessary to add the following header to the request:
```
X-Agouti-Enable: 1
```

It is possible to customize the length of the content that the server will respond with the following header:
```
X-Agouti-Limit: 14000
```

### Example usage with Cucumber, Capybara and Poltergeist
```
Given /^(?:|I )navigate to '(.+)'$/ do |page_path|
  visit page_path

Given /^(?:|I )navigate to '(.+)' waiting only one tcp round trip$/ do |page_path|
  page.driver.add_header("X-Agouti-Enable", "1", permanent: false)
  visit page_path
end
```

## Contributing

1. Fork it ( https://github.com/CWISoftware/agouti/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# Changelog #

## 0.0.1

* Main feature implemented.

## 0.0.2

* Refactoring, add tests and code documentation.

## 0.0.3

* Fix issues


spec/spec_helper.rb
require 'bundler/setup'
Bundler.setup

require 'agouti'

RSpec.configure do |config|
end

spec/rack/package_limiter_spec.rb
require 'spec_helper'

describe Agouti::Rack::PackageLimiter do

  let(:app) { ->(env) { [200, headers, []] } }

  describe '#call' do

    subject { described_class.new(app).call(env) }

    context 'when header X-Agouti-Enable is set' do

      context 'when header X-Agouti-Enable is set with 1' do
        let(:headers) { { 'X-Agouti-Enable' => '1' } }
        let(:env) { { 'HTTP_X_AGOUTI_ENABLE' => '1' } }

        context 'when the request response is not an html' do
          it 'returns status code 204' do
            expect(subject). to match_array([204, {}, []])
          end
        end

        context 'when header X-Agouti-Limit is not set' do
          it 'returns gzipped data truncated at 14000 bytes' do
            headers.merge!('Content-Type' => 'text/html')

            response_status, response_headers, response_body = subject

            expect(response_status).to eq(200)
            expect(response_headers).to include(headers.merge!('Content-Encoding' => 'gzip'))
            expect(response_body).to be_a(Agouti::Rack::PackageLimiter::GzipTruncatedStream)
            expect(response_body.instance_variable_get(:@byte_limit)).to eq(14000)
          end
        end

        context 'when header X-Agouti-Limit is set' do

          context 'when header X-Agouti-Limit is set with a valid number of bytes' do
            it 'returns gzipped data with given number of bytes' do
              headers.merge!('X-Agouti-Limit' => '10', 'Content-Type' => 'text/html')
              env.merge!('HTTP_X_AGOUTI_LIMIT' => '10')

              response_status, response_headers, response_body = subject

              expect(response_status).to eq(200)
              expect(response_headers).to include(headers.merge!('Content-Encoding' => 'gzip'))
              expect(response_body).to be_a(Agouti::Rack::PackageLimiter::GzipTruncatedStream)
              expect(response_body.instance_variable_get(:@byte_limit)).to eq(10)
            end
          end

          context 'when header X-Agouti-Limit is set with an invalid value' do
            let(:headers) { { 'X-Agouti-Limit' => 'foobar' } }
            let(:env) { { 'HTTP_X_AGOUTI_LIMIT' => 'foobar' } }

            it 'raises invalid header exception' do
              expect { subject }.to raise_exception(Agouti::Rack::PackageLimiter::InvalidHeaderException)
            end
          end
        end
      end

      context 'when header X-Agouti-Enable is set with 0' do
        let(:headers) { { 'X-Agouti-Enable' => '0' } }
        let(:env) { { 'HTTP_X_AGOUTI_ENABLE' => '0' } }

        it 'does nothing' do
          expect(subject). to match_array([200, headers, []])
        end
      end

      context 'when header X-Agouti-Enable is set with an invalid value' do
        let(:headers) { { 'X-Agouti-Enable' => 'foobar' } }
        let(:env) { { 'HTTP_X_AGOUTI_ENABLE' => 'foobar' } }

        it 'raises invalid header exception' do
          expect { subject }.to raise_exception(Agouti::Rack::PackageLimiter::InvalidHeaderException)
        end
      end
    end

    context 'when header X-Agouti-Enable is not set' do
      let(:headers) { { } }
      let(:env) { { } }

      it 'does nothing' do
        expect(subject). to match_array([200, headers, []])
      end
    end
  end
end


lib/agouti/rack/package_limiter.rb
require 'rack'

module Agouti
  module Rack

    # Public: rack middleware that truncates the gzipped response.
    # Useful for testing critical rendering path optimization.
    class PackageLimiter

      ENABLE_HEADER = 'X-Agouti-Enable'
      LIMIT_HEADER = 'X-Agouti-Limit'
      # Public: Default limit of bytes.
      DEFAULT_LIMIT = 14000

      # Public: Constructor.
      #
      # app - rack app instance
      #
      # Returns an instance of Agouti::Rack::PackageLimiter middleware.
      def initialize(app)
        @app = app
      end

      # Public: Apply middleware to request.
      #
      # env - environment.
      #
      # Raises Agouti::Rack::PackageLimiter::InvalidHeaderException if headers are not valid. The following values are accepted:
      #   X-Agouti-Enable:
      #   - header not present or set with value 0(disabled).
      #   - header set with value 1 (enabled).
      #   X-Agouti-Limit: a positive integer.
      #
      # The response body is gzipped only when the following conditions are met:
      #   Header X-Agouti-Enable set with value 1 and header Content-Type with value 'text/html'.
      #   If header X-Agouti-Limit is set, response body will be truncated to the given number of bytes.
      #   Otherwise, body will be truncated to the default limit, which is 14000 bytes.
      #
      # If header X-Agouti-Enable is enabled but header Content-Type does not have value 'text/html',
      # the middleware will return a response with status code 204 and empty body.
      #
      # If header X-Agouti-Enable has value 0 or is empty, the response will not be modified.
      def call(env)
        raise InvalidHeaderException unless valid?(env)

        status, headers, body = @app.call(env)

        set_limit(env)

        if enabled?(env)
          unless headers['Content-Type'] && headers['Content-Type'].include?('text/html')
            return [204, {}, []]
          end

          headers = ::Rack::Utils::HeaderHash.new(headers)

          headers['Content-Encoding'] = 'gzip'
          headers.delete('Content-Length')
          mtime = headers.key?('Last-Modified') ? Time.httpdate(headers['Last-Modified']) : Time.now

          [status, headers, GzipTruncatedStream.new(body, mtime, @limit)]
        else
          [status, headers, body]
        end
      end

      private

      def get_http_header env, header
        env["HTTP_#{header.upcase.gsub('-', '_')}"]
      end

      def enabled? env
        get_http_header(env, ENABLE_HEADER).to_i == 1
      end

      def set_limit env
        @limit = (get_http_header(env, LIMIT_HEADER)) ? get_http_header(env, LIMIT_HEADER).to_i : DEFAULT_LIMIT
      end

      def valid? env
        valid_enable_header?(env) && valid_limit_header?(env)
      end

      def parseable?(header)
        (header =~ /^[0-9]+$/) == 0
      end

      # Defines two methods: valid_enable_header? and valid_limit_header?
      ['enable', 'limit'].each do |action|
        define_method("valid_#{action}_header?") do |env|
          header = get_http_header(env, self.class.const_get("#{action.upcase}_HEADER"))

          return true if header.nil?

          if parseable?(header)
            header.to_i >= 0
          else
            false
          end
        end
      end

      # Public: class responsible for truncating the gzip stream to a given number of bytes.
      class GzipTruncatedStream < ::Rack::Deflater::GzipStream
        # Public: Constructor.
        #
        # body - response body.
        # mtime - last-modified time.
        # byte_limit - byte limit.
        #
        # Returns an instance of Agouti::Rack::PackageLimiter::GzipTruncatedStream.
        def initialize body, mtime, byte_limit
          super body, mtime
          @byte_limit = byte_limit
          @total_sent_bytes = 0
        end

        # Public: Writes data to stream.
        #
        # data - data.
        #
        # If total sent bytes reaches bytes limit, data is sliced.
        def write(data)
          if @total_sent_bytes + data.bytesize > @byte_limit
            data = data.byteslice(0, @byte_limit - @total_sent_bytes)
          end

          @total_sent_bytes += data.bytesize
          @writer.call(data)
        end
      end

      # Public: custom exception class for invalid headers.
      class InvalidHeaderException < Exception; end;
    end
  end
end
